plugins/egis-moc/fu-egis-moc-device.c

```diff
@@ -0,0 +1,539 @@
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, "init");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 99, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, "fini");
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	fu_progress_step_done(progress);
+
+	/* write each block */
+	chunks = fu_chunk_array_new_from_bytes(fw,
+					       FU_CHUNK_ADDR_OFFSET_NONE,
+					       FU_CHUNK_PAGESZ_NONE,
+					       FU_EGIS_MOC_FLASH_TRANSFER_BLOCK_SIZE);
+	if (!fu_egis_moc_device_write_packets(self, chunks, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	if (!fu_egis_moc_device_write_checksum(self, chunks, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success! */
+	return TRUE;
+}
+
+static gboolean
+fu_egis_moc_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuEgisMocDevice *self = FU_EGIS_MOC_DEVICE(device);
+
+	if (!fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_debug("already in runtime mode, skipping");
+		return TRUE;
+	}
+
+	/* success */
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_egis_moc_device_detach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuEgisMocDevice *self = FU_EGIS_MOC_DEVICE(device);
+
+	if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		g_debug("already in bootloader mode, skipping");
+		return TRUE;
+	}
+	// TODO
+	if (!fu_egis_moc_device_update_init(self, error)) {
+		g_prefix_error(error, "failed to detach: ");
+		return FALSE;
+	}
+
+	/* success */
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static void
+fu_egis_moc_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DECOMPRESSING, 0, "prepare-fw");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 7, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 42, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 51, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, "reload");
+}
+
+static gchar *
+fu_egis_moc_device_convert_version(FuDevice *device, guint64 version_raw)
+{
+	return fu_version_from_uint32(version_raw, fu_device_get_version_format(device));
+}
+
+static void
+fu_egis_moc_device_init(FuEgisMocDevice *self)
+{
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SELF_RECOVERY);
+	fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_USE_RUNTIME_VERSION);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_PLAIN);
```

Comment:
is this really the right version format?  Like a string that should always be compared alphanumerically?  Usually it should be something like int, hex, quad, etc.

Reply:
good catch; changed to quad.